File: app/src/main/java/vasuki/istanpdf/QuadrantCircleView.java
package vasuki.istanpdf;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.view.View;

public class QuadrantCircleView extends View {
    private final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final RectF circleBounds = new RectF();
    private ThemePrefs.Accent accent;
    private boolean selected;
    private boolean amoled;

    public QuadrantCircleView(Context context) {
        super(context);
        setLayerType(View.LAYER_TYPE_SOFTWARE, null);
    }

    public void bind(ThemePrefs.Accent accent, boolean selected, boolean amoled) {
        this.accent = accent;
        this.selected = selected;
        this.amoled = amoled;
        setSelected(selected);
        setFocusable(true);
        setContentDescription(accent.name + " accent" + (selected ? ", selected" : ""));
        invalidate();
    }

    @Override
    public void onInitializeAccessibilityNodeInfo(android.view.accessibility.AccessibilityNodeInfo info) {
        super.onInitializeAccessibilityNodeInfo(info);
        info.setClassName("android.widget.RadioButton");
        info.setCheckable(true);
        info.setChecked(selected);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (accent == null) return;

        float inset = dp(selected ? 3 : 4);
        float size = Math.min(getWidth(), getHeight()) - (2 * inset);
        float left = (getWidth() - size) / 2f;
        float top = (getHeight() - size) / 2f;
        circleBounds.set(left, top, left + size, top + size);
        int[] colors = ThemePrefs.quadrantColors(accent, amoled);
        for (int i = 0; i < colors.length; i++) {
            paint.setStyle(Paint.Style.FILL);
            paint.setColor(colors[i]);
            canvas.drawArc(circleBounds, -90 + (i * 90), 90, true, paint);
        }

        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(dp(selected ? 2 : 1));
        paint.setColor(selected ? ThemePrefs.accentForeground(accent, amoled)
                : ThemePrefs.tint(accent.dark, amoled ? 0xFF000000 : 0xFFFFFFFF, amoled ? 0.2f : 0.58f));
        canvas.drawOval(circleBounds, paint);

    }

    private int dp(int value) {
        return Math.round(value * getResources().getDisplayMetrics().density);
    }
}

File: app/src/main/kotlin/vasuki/istanpdf/QuadrantCircleView.kt
package vasuki.istanpdf

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.view.View
import android.view.accessibility.AccessibilityNodeInfo
import kotlin.math.min
import kotlin.math.roundToInt

class QuadrantCircleView(context: Context) : View(context) {
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val circleBounds = RectF()
    private var accent: ThemePrefs.Accent? = null
    private var selected = false
    private var amoled = false

    init {
        setLayerType(LAYER_TYPE_SOFTWARE, null)
    }

    fun bind(accent: ThemePrefs.Accent, selected: Boolean, amoled: Boolean) {
        this.accent = accent
        this.selected = selected
        this.amoled = amoled
        isSelected = selected
        isFocusable = true
        contentDescription = accent.name + " accent" + if (selected) ", selected" else ""
        invalidate()
    }

    override fun onInitializeAccessibilityNodeInfo(info: AccessibilityNodeInfo) {
        super.onInitializeAccessibilityNodeInfo(info)
        info.className = "android.widget.RadioButton"
        info.isCheckable = true
        info.isChecked = selected
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val accent = accent ?: return

        val inset = dp(if (selected) 3 else 4)
        val size = min(width, height) - (2 * inset)
        val left = (width - size) / 2f
        val top = (height - size) / 2f
        circleBounds.set(left, top, left + size, top + size)
        val colors = ThemePrefs.quadrantColors(accent, amoled)
        for (i in colors.indices) {
            paint.style = Paint.Style.FILL
            paint.color = colors[i]
            canvas.drawArc(circleBounds, -90f + (i * 90), if (amoled) 90.5f else 90f, true, paint)
        }

        val strokeColor = if (selected) ThemePrefs.accentForeground(accent, amoled)
        else ThemePrefs.tint(accent.dark, if (amoled) 0xFF000000.toInt() else 0xFFFFFFFF.toInt(), if (amoled) 0.2f else 0.58f)

        paint.style = Paint.Style.STROKE
        paint.color = strokeColor

        if (!amoled) {
            paint.strokeWidth = dp(1).toFloat()
            val cx = circleBounds.centerX()
            val cy = circleBounds.centerY()
            canvas.drawLine(cx, circleBounds.top, cx, circleBounds.bottom, paint)
            canvas.drawLine(circleBounds.left, cy, circleBounds.right, cy, paint)
        }

        paint.strokeWidth = dp(if (selected) 2 else 1).toFloat()
        canvas.drawOval(circleBounds, paint)
    }

    private fun dp(value: Int): Float =
        Math.round(value * resources.displayMetrics.density).toFloat()
}
